File: go/record_modes_conformance_test.go
package cassetter_test

import (
	"bytes"
	"errors"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"reflect"
	"sort"
	"testing"

	"github.com/Kludex/cassetter/go"
)

type storedInteraction struct {
	URI    string `json:"uri"`
	Status int    `json:"status"`
}

type recordModeCase struct {
	Name              string               `json:"name"`
	Mode              cassetter.RecordMode `json:"mode"`
	Existing          bool                 `json:"existing"`
	Requests          []string             `json:"requests"`
	ExpectedOutcomes  []string             `json:"expectedOutcomes"`
	ExpectedBaseCalls int                  `json:"expectedBaseCalls"`
	ExpectedFile      *[]storedInteraction `json:"expectedFile"`
}

func TestRecordModeConformance(t *testing.T) {
	t.Parallel()
	fixtures := filepath.Join(conformanceFixtures(t), "record-modes")
	cases := loadJSONFixture[[]recordModeCase](t, filepath.Join(fixtures, "cases.json"))
	for _, testCase := range cases {
		t.Run(testCase.Name, func(t *testing.T) {
			t.Parallel()
			path := filepath.Join(t.TempDir(), "cassette.yaml")
			if testCase.Existing {
				content, err := os.ReadFile(filepath.Join(fixtures, "existing.yaml"))
				if err != nil {
					t.Fatal(err)
				}
				if err := os.WriteFile(path, content, 0o600); err != nil {
					t.Fatal(err)
				}
			}
			baseCalls := 0
			base := roundTripFunc(func(request *http.Request) (*http.Response, error) {
				baseCalls++
				content := []byte(`{"source":"live"}`)
				return &http.Response{
					StatusCode:    299,
					Header:        http.Header{"content-type": []string{"application/json"}},
					Body:          io.NopCloser(bytes.NewReader(content)),
					ContentLength: int64(len(content)),
					Request:       request,
				}, nil
			})
			transport := cassetter.NewTransport(
				base,
				cassetter.WithPath(path),
				cassetter.WithRecordMode(testCase.Mode),
			)
			if err := transport.Initialize(); err != nil {
				t.Fatal(err)
			}
			client := &http.Client{Transport: transport}
			outcomes := make([]string, 0, len(testCase.Requests))
			for _, uri := range testCase.Requests {
				callsBefore := baseCalls
				response, err := client.Get(uri)
				if errors.Is(err, cassetter.ErrNoMatch) {
					outcomes = append(outcomes, "no_match")
					continue
				}
				if err != nil {
					t.Fatal(err)
				}
				if baseCalls == callsBefore {
					outcomes = append(outcomes, "replay")
				} else {
					outcomes = append(outcomes, "live")
				}
				if _, err := io.Copy(io.Discard, response.Body); err != nil {
					t.Fatal(err)
				}
				if err := response.Body.Close(); err != nil {
					t.Fatal(err)
				}
			}
			if err := transport.Close(); err != nil {
				t.Fatal(err)
			}
			if !reflect.DeepEqual(outcomes, testCase.ExpectedOutcomes) {
				t.Fatalf("outcomes = %v, want %v", outcomes, testCase.ExpectedOutcomes)
			}
			if baseCalls != testCase.ExpectedBaseCalls {
				t.Fatalf("base calls = %d, want %d", baseCalls, testCase.ExpectedBaseCalls)
			}
			assertStoredInteractions(t, path, testCase.ExpectedFile)
		})
	}
}

func assertStoredInteractions(t *testing.T, path string, expected *[]storedInteraction) {
	t.Helper()
	if expected == nil {
		if _, err := os.Stat(path); !errors.Is(err, os.ErrNotExist) {
			t.Fatalf("cassette exists or returned unexpected error: %v", err)
		}
		return
	}
	cassette, err := cassetter.Load(path)
	if err != nil {
		t.Fatal(err)
	}
	actual := make([]storedInteraction, 0, len(cassette.Interactions))
	for _, interaction := range cassette.Interactions {
		actual = append(actual, storedInteraction{
			URI:    interaction.Request.URI,
			Status: interaction.Response.Status,
		})
	}
	sort.Slice(actual, func(left int, right int) bool {
		return actual[left].URI < actual[right].URI
	})
	if !reflect.DeepEqual(actual, *expected) {
		t.Fatalf("stored interactions = %v, want %v", actual, *expected)
	}
}

File: ts/tests/record-modes-conformance.test.ts
import {
  copyFileSync,
  existsSync,
  mkdtempSync,
  readFileSync,
  rmSync,
} from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { expect, it } from "vitest";

import { native } from "../src/binding.js";
import { Cassette, NoMatchError } from "../src/cassette.js";
import { RecordMode } from "../src/recording.js";
import { CONFORMANCE_ROOT } from "./conformance-helpers.js";

const FIXTURES = join(CONFORMANCE_ROOT, "conformance", "record-modes");

type StoredInteraction = {
  uri: string;
  status: number;
};

type RecordModeCase = {
  name: string;
  mode: RecordMode;
  existing: boolean;
  requests: string[];
  expectedOutcomes: string[];
  expectedBaseCalls: number;
  expectedFile: StoredInteraction[] | null;
};

const CASES = JSON.parse(
  readFileSync(join(FIXTURES, "cases.json"), "utf-8"),
) as RecordModeCase[];

it.each(CASES)("applies shared record mode case: $name", (case_) => {
  const directory = mkdtempSync(join(tmpdir(), "cassetter-modes-"));
  const path = join(directory, "cassette.yaml");
  try {
    if (case_.existing) {
      copyFileSync(join(FIXTURES, "existing.yaml"), path);
    }
    const cassette = new Cassette(path, { recordMode: case_.mode });
    cassette.load();
    const outcomes: string[] = [];
    let baseCalls = 0;

    for (const uri of case_.requests) {
      try {
        cassette.play("GET", uri, {}, null);
        outcomes.push("replay");
      } catch (error) {
        if (!(error instanceof NoMatchError)) {
          throw error;
        }
        if (!cassette.canRecord) {
          outcomes.push("no_match");
          continue;
        }
        cassette.record(
          "GET",
          uri,
          {},
          null,
          299,
          { "content-type": ["application/json"] },
          Buffer.from('{"source":"live"}'),
        );
        outcomes.push("live");
        baseCalls += 1;
      }
    }
    cassette.save();

    expect(outcomes).toEqual(case_.expectedOutcomes);
    expect(baseCalls).toBe(case_.expectedBaseCalls);
    if (case_.expectedFile === null) {
      expect(existsSync(path)).toBe(false);
    } else {
      const stored = native.Cassette.load(path);
      const actual = stored.interactions
        .map((interaction) => ({
          uri: interaction.request.uri,
          status: interaction.response.status,
        }))
        .sort((left, right) => left.uri.localeCompare(right.uri));
      expect(actual).toEqual(case_.expectedFile);
    }
  } finally {
    rmSync(directory, { recursive: true, force: true });
  }
});
